code/modules/antagonists/wizard/equipment/wizard_spellbook.dm

```diff
@@ -0,0 +1,329 @@
+		if(contract.used)
+			to_chat(user, ("<span class='warning'>The contract has been used, you can't get your points back now!</span>"))
+			return
+
+		to_chat(user, ("<span class='notice'>You feed the contract back into the spellbook, refunding your points.</span>"))
+		uses += contract_entry.cost
+		contract_entry.times--
+		qdel(O)
+
+	else if(istype(O, /obj/item/antag_spawner/slaughter_demon/laughter))
+		var/datum/spellbook_entry/item/hugbottle/demon_entry = locate() in entries
+		if(!istype(demon_entry))
+			to_chat(user, ("<span class='warning'>[src] doesn't seem to want to refund [O].</span>"))
+			return
+		if(!demon_entry.can_refund(user, src))
+			to_chat(user, ("<span class='warning'>You can't refund [O].</span>"))
+			return
+
+		to_chat(user, ("<span class='notice'>On second thought, maybe summoning a demon isn't a funny idea. You refund your points.</span>"))
+		uses += demon_entry.cost
+		demon_entry.times--
+		qdel(O)
+
+	else if(istype(O, /obj/item/antag_spawner/slaughter_demon))
+		var/datum/spellbook_entry/item/bloodbottle/demon_entry = locate() in entries
+		if(!istype(demon_entry))
+			to_chat(user, ("<span class='warning'>[src] doesn't seem to want to refund [O].</span>"))
+			return
+		if(!demon_entry.can_refund(user, src))
+			to_chat(user, ("<span class='warning'>You can't refund [O].</span>"))
+			return
+
+		to_chat(user, ("<span class='notice'>On second thought, maybe summoning a demon is a bad idea. You refund your points.</span>"))
+		uses += demon_entry.cost
+		demon_entry.times--
+		qdel(O)
+
+	return ..()
+
+/// Instantiates our list of spellbook entries.
+/obj/item/spellbook/proc/prepare_spells()
+	var/entry_types = subtypesof(/datum/spellbook_entry)
+	for(var/type in entry_types)
+		var/datum/spellbook_entry/possible_entry = new type()
+		if(!possible_entry.can_be_purchased())
+			qdel(possible_entry)
+			continue
+
+		possible_entry.set_spell_info() //loads up things for the entry that require checking spell instance.
+		entries |= possible_entry
+
+/obj/item/spellbook/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Spellbook")
+		ui.open()
+
+/obj/item/spellbook/ui_data(mob/user)
+	var/list/data = list()
+	data["owner"] = owner
+	data["points"] = uses
+	data["semi_random_bonus"] = initial(uses) + semi_random_bonus
+	data["full_random_bonus"] = initial(uses) + full_random_bonus
+	return data
+
+//This is a MASSIVE amount of data, please be careful if you remove it from static.
+/obj/item/spellbook/ui_static_data(mob/user)
+	var/list/data = list()
+	// Collect all info from each intry.
+	var/list/entry_data = list()
+	for(var/datum/spellbook_entry/entry as anything in entries)
+		var/list/individual_entry_data = list()
+		individual_entry_data["name"] = entry.name
+		individual_entry_data["desc"] = entry.desc
+		individual_entry_data["ref"] = REF(entry)
+		individual_entry_data["requires_wizard_garb"] = entry.requires_wizard_garb
```

Comment:
`requires_wizard_garb` is `clothes_req` in the newer tgui file so would also need to be changed there as well